gocd/bci_repo_publish.py

```diff
@@ -0,0 +1,199 @@
+        return mtime[0]
+
+    def openqa_jobs_for_product(self, arch, version, build):
+        """Query openQA for all relevant jobs"""
+        values = {
+            'group': 'BCI repo',
+            'flavor': 'BCI-Repo-Updates',
+            'arch': arch,
+            'version': version,
+            'build': build,
+            'scope': 'current',
+            'latest': '1',
+        }
+        return self.openqa.openqa_request('GET', 'jobs', values)['jobs']
+
+    def is_repo_published(self, project, repo):
+        """Validates that the given prj/repo is fully published and all builds
+        have succeeded."""
+        url = makeurl(self.apiurl, ['build', project, '_result'],
+                      {'view': 'summary', 'repository': repo})
+        root = ET.parse(http_GET(url)).getroot()
+        for result in root.findall('result'):
+            if result.get('dirty', 'false') != 'false':
+                return False
+            if result.get('code') != 'published' or result.get('state') != 'published':
+                return False
+
+        for statuscount in root.findall('statuscount'):
+            if statuscount.get('code') not in ['succeeded', 'disabled', 'excluded']:
+                return False
+
+        return True
+
+    def run(self, version, token=None):
+        build_prj = f'SUSE:SLE-{version}:Update:BCI'
+
+        if not self.is_repo_published(build_prj, 'images'):
+            self.logger.info(f'{build_prj}/images not successfully built')
+            return
+
+        # Build the list of packages with metainfo
+        packages = []
+        # As long as it's the same everywhere, hardcoding this list here
+        # is easier and safer than trying to derive it from the package list.
+        for arch in ['aarch64', 'ppc64le', 's390x', 'x86_64']:
+            packages.append({
+                'arch': arch,
+                'name': f'000product:SLE_BCI-ftp-POOL-{arch}',
+                'build_prj': build_prj,
+                'publish_prj': f'SUSE:Products:SLE-BCI:{version}:{arch}'
+            })
+
+        # Fetch the build numbers of built products.
+        # After release, the BuildXXX part vanishes, so the mtime has to be
+        # used instead for comparing built and published binaries.
+        for pkg in packages:
+            pkg['built_version'] = self.version_of_product(pkg['build_prj'], pkg['name'],
+                                                           'images', 'local')
+            pkg['built_mtime'] = self.mtime_of_product(pkg['build_prj'], pkg['name'],
+                                                       'images', 'local')
+            pkg['published_mtime'] = self.mtime_of_product(pkg['publish_prj'], pkg['name'],
+                                                           'images', 'local')
+
+        # Verify that the builds for all archs are in sync
+        built_versions = set([pkg['built_version'] for pkg in packages])
```

Comment:
```suggestion
        built_versions = {pkg['built_version'] for pkg in packages}
```

Reply:
What's this syntax? Never seen it before

Reply:
https://python-reference.readthedocs.io/en/latest/docs/comprehensions/set_comprehension.html

Reply:
Does not work:

```
  File "/home/fvogt/osc-plugin-factory/gocd/./bci_repo_publish.py", line 110, in run
    built_versions = {[pkg['built_version'] for pkg in packages]}
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unhashable type: 'list'
```